### ISSUE TITLE
[prefer-optional-chain] Suggest `(foo || {}).bar` as a valid match for the rule

### ISSUE THREAD

**Issue by omril1** [typescript-eslint/typescript-eslint#4395]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-optional-chain": "error",
  }
}
```

```TS
const foo = (bar || {}).baz;
// or
const foo = (bar|| {})[baz];
// or
const foo = (((a || {}).b || {}).c || {}).d;
```

**Expected Result**
The rule should suggest to use optional chaining

I'm working on a codebase that uses this pattern a lot (about ~60% of the cases I refactor to optional chaining arise from this pattern, the others come from `&&`s) and I'm only getting prompted for the `&&` cases.

I expect to get the same warnings/errors/suggestions the `&&` case have, not including function calls

**Actual Result**
No error or warning

**Additional Info**
Willing to add a PR if needed
Tested this already against existing code in this repo

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.9.0` |
| `@typescript-eslint/parser`        | `5.9.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.1.0` |
| `node`                             | `14.17.5` |

**Comment by glen-84** [comment]
I apologize for commenting on a closed issue, but I didn't want to open a new one if I may be missing something obvious.

Optional chaining only handles `null` and `undefined`, whereas logical OR handles any falsy value, which means that the suggested fix can result in invalid code.

[Example](https://typescript-eslint.io/play#ts=4.6.2&sourceType=module&code=MYewdgzgLgBAZiEMC8MCyBDKALAdAJwzABMQBbACgEoYB+eDAGwgFMYAuGAbwCMN9OARgBMAZgC+AbgCwAKDkB6BTBYRGASzBQAtMXUQMPRi21gWADx0azMbdpgA5JABUAyivz4Q+CLjmhIWHMUGAoEJAAfCO5xKlw+fBl5WSUYADF1YIwAB2yNFmJcGDcPL3w-WQDoGABPEPDaeP5JIA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9DCnAGZwpYD2matBAhrlkgBbOEgBcI1FLRQgAvkA&tsConfig=N4XyA)

**Comment by bradzacher** [comment]
This is a known issue with the rule in general, and is not limited to just this one case.

For example:
```ts
declare const x: {foo: false | {bar: string}};

x.foo || x.foo.bar
```

The rule will error and suggest fix this, but it is wrong.

Suggestion fixers are allowed to be wrong though. That's the point of having them as opposed to an automatic fixer - which must always be correct.

This is just a known weakness in the rule, unfortunately, as it doesn't use type information.

**Comment by omril1** [comment]
@glen-84 This weakness in the rule also exists in the `&&` suggestion, and it's documented as a note.
https://github.com/typescript-eslint/typescript-eslint/blob/d21cfe0f4b7d3041948b1b6e0cd56c5ec34b2b3f/packages/eslint-plugin/docs/rules/prefer-optional-chain.md#L86

**Comment by glen-84** [comment]
Ah, sorry for missing that, and thanks for the explanation. 👍
